### PR TITLE
Added Azure.VM.MaintenanceConfig

### LINKED ISSUE #2121: Use a VM maintenance configuration for virtual machines (preview)
### Existing rule

_No response_

### Suggested rule

VMs can be assigned a maintenance configuration which allows managed assessments and updates for VM patches within the guest operating system.

This can be done by creating a `Microsoft.Maintenance/configurationAssignments` sub resource with the linked maintenance configuration resource Id.

For in-flight resources we need to collect and export this data.

### Pillar

Operational Excellence

### Additional context

- https://learn.microsoft.com/azure/templates/microsoft.maintenance/configurationassignments
- https://learn.microsoft.com/azure/update-center/manage-vms-programmatically
- https://learn.microsoft.com/azure/update-center/manage-update-settings

### PR DESCRIPTION
**Pull request by BenjaminEngeset**
## PR Summary

Fixes #2121

Added Azure.VM.MaintenanceConfig rule.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section
- **Other code changes**
  - [ ] Unit tests created/ updated
  - [ ] Link to a filed issue
  - [ ] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

**Comment by BenjaminEngeset** [comment]
@BernieWhite Ready for review. No pre-condition for operating systems here either.

File: docs/en/rules/Azure.VM.MaintenanceConfig.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  scope: vm
+  properties: {
+    maintenanceConfigurationId: maintenanceConfigurationId
+  }
+}
+```
+
+## NOTES
+
+Not all operating systems are supported, check out the `LINKS` section for more information. Update management center doesn't support driver updates.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.VM.MaintenanceConfig.md:72]:
```suggestion
Operating system updates with Update Managment center is a preview feature.
Not all regions or operating systems are supported, check out the `LINKS` section for supported regions.
Update management center doesn't support driver updates.
```

File: docs/en/rules/Azure.VM.MaintenanceConfig.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Use a maintenance configuration for virtual machines.
+
+## DESCRIPTION
+
+Virtual machines can be attached to a maintenance configuration which allows customer managed assessments and updates for machine patches within the guest operating system.
+
+## RECOMMENDATION
+
+Consider associating a maintenance configuration for virtual machines.
```
Comment (BernieWhite) [inline docs/en/rules/Azure.VM.MaintenanceConfig.md:21]:
```suggestion
Consider automatically managing and applying operating system updates by associating a maintenance configuration.
```

File: docs/en/rules/Azure.VM.MaintenanceConfig.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+```json
+{
+  "type": "Microsoft.Maintenance/configurationAssignments",
+  "apiVersion": "2022-11-01-preview",
+  "name": "[parameters('assignmentName')]",
+  "location": "[parameters('location')]",
+  "scope": "[format('Microsoft.Compute/virtualMachines/{0}', parameters('name'))]",
+  "properties": {
+    "maintenanceConfigurationId": "[format('/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/test-rg/providers/Microsoft.Maintenance/maintenanceConfigurations/{0}', parameters('maintenanceConfigName'))]",
```
Comment (BernieWhite) [inline docs/en/rules/Azure.VM.MaintenanceConfig.md:42]:
```suggestion
    "maintenanceConfigurationId": "[parameters('maintenanceConfigurationId')]"
```

File: docs/en/rules/Azure.VM.MaintenanceConfig.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+To deploy virtual machines that pass this rule:
+
+- Deploy a `Microsoft.Maintenance/configurationAssignments` sub-resource (extension resource).
+- Set the `properties.maintenanceConfigurationId` property to the linked maintenance configuration resource Id.
+
+For example:
+
+```bicep
+resource assosciateMaintenanceConfig 'Microsoft.Maintenance/configurationAssignments@2022-11-01-preview' = {
```
Comment (BernieWhite) [inline docs/en/rules/Azure.VM.MaintenanceConfig.md:60]:
```suggestion
resource config 'Microsoft.Maintenance/configurationAssignments@2022-11-01-preview' = {
```

**Review by BernieWhite — CHANGES_REQUESTED** [review]
Nice work @BenjaminEngeset. I've added a few suggestions.

Let me know if you have any questions.

File: docs/en/rules/Azure.VM.MaintenanceConfig.md (hunk 1 above)
Comment (BenjaminEngeset) [reply docs/en/rules/Azure.VM.MaintenanceConfig.md:72]: @BernieWhite Minor feedback from me, for Azure VMs all regions are supported, compared to Arc-enabled servers.

`Update management center (preview) is available in all Azure public regions where compute virtual machines are available.`

https://learn.microsoft.com/azure/update-center/support-matrix?tabs=azurevm%2Cazurevm-os#supported-regions

Should we add this anyways?

Comment (BernieWhite) [reply docs/en/rules/Azure.VM.MaintenanceConfig.md:72]: Good point @BenjaminEngeset. We should still add the preview comment, but no comment about region limiting.

For example:

> Operating system updates with Update Managment center is a preview feature.
> Update management center doesn't support driver updates.

Comment (BenjaminEngeset) [reply docs/en/rules/Azure.VM.MaintenanceConfig.md]: @BernieWhite Check latest update from me now and see if you think this is suitable.

Comment (BernieWhite) [reply docs/en/rules/Azure.VM.MaintenanceConfig.md]: @BenjaminEngeset Looks good to me.

**Review by BernieWhite — APPROVED** [review]
Nice work @BenjaminEngeset. All good to merge.

**Pull request merged**